src/metabase/lib/join.cljc

```diff
@@ -37,44 +23,58 @@
    [metabase.lib.util :as lib.util]
    [metabase.mbql.util.match :as mbql.u.match]
    [metabase.shared.util.i18n :as i18n]
+   [metabase.util :as u]
    [metabase.util.malli :as mu]))
 
-(defmulti with-join-alias-method
-  "Implementation for [[with-join-alias]]."
-  {:arglists '([field-or-joinable join-alias])}
-  (fn [field-or-joinable _join-alias]
-    (lib.dispatch/dispatch-value field-or-joinable)))
-
-(defmethod with-join-alias-method :dispatch-type/fn
-  [f join-alias]
-  (fn [query stage-number]
-    (let [field-or-joinable (f query stage-number)]
-      (with-join-alias-method field-or-joinable join-alias))))
-
-(defmethod with-join-alias-method :mbql/join
-  [join join-alias]
-  (assoc join :alias join-alias))
+(def ^:private JoinWithOptionalAlias
+  "A Join that may not yet have an `:alias`, which is normally required; [[join]] accepts this and will add a default
+  alias if one is not present."
+  [:merge
+   [:ref ::lib.schema.join/join]
+   [:map
+    [:alias {:optional true} [:ref ::lib.schema.join/alias]]]])
+
+(def ^:private PartialJoin
+  "A join that may not yet have an `:alias` or `:conditions`."
+  [:merge
+   JoinWithOptionalAlias
+   [:map
+    [:conditions {:optional true} [:ref ::lib.schema.join/conditions]]]])
+
+(def ^:private FieldOrPartialJoin
+  [:or
+   lib.metadata/ColumnMetadata
+   [:ref :mbql.clause/field]
+   PartialJoin])
 
 (mu/defn with-join-alias
-  "Add a specific `join-alias` to `field-or-joinable`, with is either a `:field`/Field metadata, or something 'joinable'
-  like a join map or Table metadata. Does not recursively update other references (yet; we can add this in the
-  future)."
-  [field-or-joinable join-alias :- ::lib.schema.common/non-blank-string]
-  (with-join-alias-method field-or-joinable join-alias))
-
-(defmulti current-join-alias-method
-  "Impl for [[current-join-alias]]."
-  {:arglists '([field-or-joinable])}
-  lib.dispatch/dispatch-value)
-
-(defmethod current-join-alias-method :default
-  [_x]
-  nil)
+  "Add OR REMOVE a specific `join-alias` to `field-or-joinable`, with is either a `:field`/Field metadata, or something
+  'joinable' like a join map or Table metadata. Does not recursively update other references (yet; we can add this in
+  the future)."
+  {:style/indent [:form]}
+  [field-or-joinable :- [:or FieldOrPartialJoin fn?]
```

Comment:
I'm not really sure I understand what change you're suggesting... this was a multimethod before this PR, I changed it to just be a plain function. Are you suggesting that we split this out into separate functions for the join alias for a Join vs the join alias for a Column? I guess that could make sense

Reply:
Yes, that was my "suggestion". I was also thinking "out loud" if we want to make similar changes to some of the other multimethods (i.e., making them simple functions and potentially breaking them up), but I don't expect this from this PR.

Reply:
Yeah, I'm not against that change at all, but I think doing it in a follow-on would be better for me since I already have other PRs based off of this one. I think we definitely should spend some time cleaning up the API more once we wrap more stuff up